sa/storage-authority.go

```diff
@@ -673,6 +673,117 @@ func (ssa *SQLStorageAuthority) FinalizeAuthorization(authz core.Authorization)
 	return
 }
 
+// RevokeAuthorizationsByDomain invalidates all pending or finalized authorization
+// and any associated challenges for a single domain
+func (ssa *SQLStorageAuthority) RevokeAuthorizationsByDomain(ident core.AcmeIdentifier) (int64, int64, error) {
+	invalidStatus, revokedStatus := string(core.StatusInvalid), string(core.StatusRevoked)
+	identifier, err := json.Marshal(ident)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	tx, err := ssa.dbMap.Begin()
+	if err != nil {
+		return 0, 0, err
+	}
+
+	// collect authorization IDs before beginning revocations
+	now := ssa.clk.Now()
+	var allIDs []string
+	_, err = tx.Select(
+		&allIDs,
+		`SELECT id FROM authz
+    WHERE identifier = :ident AND
+    status != :invalid AND
+    status != :revoked AND
+    expires > :now`,
+		map[string]interface{}{"ident": identifier, "invalid": invalidStatus, "revoked": revokedStatus, "now": now},
+	)
+	if err != nil {
+		tx.Rollback()
+		return 0, 0, err
+	}
+	_, err = tx.Select(
+		&allIDs,
+		`SELECT id FROM pendingAuthorizations
```

Comment:
De-dupe these two select statements by having a single string with a %s for `pendingAuthorizations` vs `authz`.